File: packages/backend/src/app/routers/rewards.py
"""Rewards list, create, redeem."""

from fastapi import APIRouter, Depends

from app.deps import get_tenant_id
from app.models.rewards import RewardCreate, RewardResponse, RewardListResponse, RedeemRequest, RedeemResponse
from app.services import rewards as svc

router = APIRouter(prefix="/programs", tags=["Rewards"])


@router.get("/{program_id}/rewards", response_model=RewardListResponse)
def list_rewards(program_id: str, tenant_id: str = Depends(get_tenant_id)):
    items = svc.list_rewards(tenant_id, program_id)
    return RewardListResponse(rewards=[RewardResponse(**x) for x in items])


@router.post("/{program_id}/rewards", response_model=dict, status_code=201)
def create_reward(program_id: str, body: RewardCreate, tenant_id: str = Depends(get_tenant_id)):
    return svc.create_reward(tenant_id, program_id, body.name, body.pointsCost)


@router.post("/{program_id}/redeem", response_model=RedeemResponse)
def redeem(program_id: str, body: RedeemRequest, tenant_id: str = Depends(get_tenant_id)):
    return svc.redeem(tenant_id, program_id, body.memberId, body.rewardId)

File: packages/web/src/api/rewards.ts
import { apiGet, apiPost } from './client'

export interface Reward {
  rewardId: string
  name: string
  pointsCost: number
  tierEligibility?: string | null
}

export interface ListRewardsResponse {
  rewards: Reward[]
}

export function listRewards(
  tenantId: string,
  programId: string,
  idToken?: string | null
): Promise<ListRewardsResponse> {
  return apiGet<ListRewardsResponse>(
    `/api/v1/programs/${programId}/rewards`,
    tenantId,
    idToken
  )
}

export function createReward(
  tenantId: string,
  programId: string,
  body: { name: string; pointsCost: number },
  idToken?: string | null
): Promise<{ rewardId: string; name: string; pointsCost: number }> {
  return apiPost(
    `/api/v1/programs/${programId}/rewards`,
    tenantId,
    body,
    idToken
  )
}

export interface RedeemResponse {
  transactionId: string
  rewardId: string
  pointsDeducted: number
  balance: number
}

export function redeem(
  tenantId: string,
  programId: string,
  body: { memberId: string; rewardId: string },
  idToken?: string | null
): Promise<RedeemResponse> {
  return apiPost<RedeemResponse>(
    `/api/v1/programs/${programId}/redeem`,
    tenantId,
    body,
    idToken
  )
}
